src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp

```diff
@@ -0,0 +1,397 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <memory>
+
+#include "DataStructures/Index.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+namespace intrp {
+
+/*!
+ * \brief Performs linear interpolation in arbitrary dimensions.
+ * The class is non-owning and expects a C-ordered array
+ *
+ *  \tparam Dimension dimensionality of the table
+ *  \tparam NumberOfVariables number of variables stored in the table
+ *  \tparam UniformSpacing indicated whether the table has uniform spacing or
+ * not. This is useful for performance reasons in finding the appropriate table
+ * index. Note that the variable index is contiguous in memory.
+ *
+ *
+ */
+
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+class MultiLinearSpanInterpolation {
+ private:
+  /// Allow extrapolation above table bounds.
+  std::array<bool, Dimension> allow_extrapolation_below_data;
+  /// Allow extrapolation below table bounds.
+  std::array<bool, Dimension> allow_extrapolation_above_data;
+  /// Spacing of the table. Only used for uniform grids
+  std::array<double, Dimension> inverseSpacing, spacing;
+  /// Number of points per dimension
+  Index<Dimension> numberOfPoints;
+
+  using DataPointer = gsl::span<double const>;
+  /// X values of the table. Only used if allocated
+  std::array<DataPointer, Dimension> x;
+  /// Y values of the table. Only used if allocated
+  DataPointer y;
+
+ public:
+  template <size_t thisDimension>
+  struct weight_t {
+    std::array<
+        double,
+        brigand::power<std::integral_constant<size_t, 2>,
+                       std::integral_constant<size_t, thisDimension>>::value>
+        weights;
+    Index<brigand::power<std::integral_constant<size_t, 2>,
+                         std::integral_constant<size_t, thisDimension>>::value>
+        index;
+  };
+
+  size_t extents(size_t const whichDimension) const {
+    return numberOfPoints[whichDimension];
+  }
+
+  void extrapolateAboveData(size_t whichDimension, bool value) {
```

Comment:
Should we have users set these in the constructor rather than later? I'm not sure of the EOS use-case, but I personally feel more comfortable if people have to choose upon construction whether or not they want extrapolation. What do you think? Maybe a `const std::array<bool, Dimension>& extrapolate_in_dimension`?

Reply:
Hmm... I mean right now this is only checked with an ASSERT, and we have as default behavior to extrapolate. If we elevate this to the constructor, I feel something else should happen in this case...